### PR TITLE
modify_S2639_layc_js_only

### PR DESCRIPTION
**Pull request by cynthiabethea**
Made some updates to rule text for Java, Javascript, and the general description for the other languages.

## Review

A dedicated reviewer checked the rule description successfully for:

- [ ] logical errors and incorrect information
- [ ] information gaps and missing content
- [ ] text style and tone

### CODE REVIEW

File: rules/S2639/javascript/rule.adoc [hunk 1]
```diff
@@ -1,8 +1,8 @@
 == Why is this an issue?
 
-Empty character classes in regular expressions don't match anything.
+Character classes in regular expressions allow you to define sets of characters for matching. An empty character class will not match any character because it has no characters within the brackets. 
 
-This is likely a sign that the regular expression does not work as intended.
+It is likely the regular expression does not work as you intended. It's good practice to avoid using empty character classes in regular expressions, as they do not serve any useful purpose in pattern matching.
```
Comment (vilchik-elena) [inline rules/S2639/javascript/rule.adoc:5]: This is a bug-detection rule, so I think saying `good practice` is not appropriate

File: rules/S2639/java/rule.adoc [hunk 2]
```diff
@@ -1,7 +1,6 @@
 == Why is this an issue?
```
Comment (vilchik-elena) [inline rules/S2639/java/rule.adoc:1]: Remove `=== Noncompliant code example`

File: rules/S2639/javascript/rule.adoc [hunk 3]
```diff
@@ -1,8 +1,8 @@
...
 
 === Noncompliant code example
```
Comment (vilchik-elena) [inline rules/S2639/javascript/rule.adoc:7]: Drop this title

**Review by vilchik-elena — CHANGES_REQUESTED** [review]
The top level description is not used anywhere, currently it holds only super-java-specific content. I suggest to drop the top level description

File: rules/S2639/rule.adoc
Comment (cynthiabethea) [inline rules/S2639/rule.adoc:1]: @vilchik-elena  I noticed this rule is on the list Fred posted, this adoc file refers to languages it was never implemented in, so I am skipping any more changes to this file and just do the changes to the Java/Javascript versions of the rules

File: rules/S2639/java/rule.adoc [hunk 4]
```diff
@@ -1,29 +1,40 @@
...
 String str = "/File|Name.txt";
-
 String clean = str.replaceAll(".",""); // Noncompliant; probably meant to remove only dot chars, but returns an empty string
-String clean2 = str.replaceAll("|","_"); // Noncompliant; yields _/_F_i_l_e_|_N_a_m_e_._t_x_t_
-String clean3 = str.replaceAll(File.separator,""); // Noncompliant; exception on Windows
-
 String clean4 = str.replaceFirst(".",""); // Noncompliant;
+----
+
+``|`` is the OR operator.  When you use ``|`` as a stand-alone, it matches the space between characters. To use it as a literal character in a regular expression, it needs to be escaped with a backslash ``\|``.
```
Comment (vilchik-elena) [inline rules/S2639/java/rule.adoc:16]:
```suggestion
* ``|`` is the OR operator.  When you use ``|`` as a stand-alone, it matches the space between characters. To use it as a literal character in a regular expression, it needs to be escaped with a backslash ``\|``.
```

File: rules/S2639/java/rule.adoc [hunk 5]
```diff
@@ -1,29 +1,40 @@
...
 
 [source,java]
 ----
 String str = "/File|Name.txt";
-
 String clean = str.replaceAll(".",""); // Noncompliant; probably meant to remove only dot chars, but returns an empty string
-String clean2 = str.replaceAll("|","_"); // Noncompliant; yields _/_F_i_l_e_|_N_a_m_e_._t_x_t_
-String clean3 = str.replaceAll(File.separator,""); // Noncompliant; exception on Windows
-
 String clean4 = str.replaceFirst(".",""); // Noncompliant;
```
Comment (vilchik-elena) [inline rules/S2639/java/rule.adoc:13]:
```suggestion
String clean2 = str.replaceFirst(".",""); // Noncompliant;
```

File: rules/S2639/java/rule.adoc [hunk 6]
```diff
@@ -1,29 +1,40 @@
...
+----
+
+``|`` is the OR operator.  When you use ``|`` as a stand-alone, it matches the space between characters. To use it as a literal character in a regular expression, it needs to be escaped with a backslash ``\|``.
+
+[source,java]
+----
+String clean2 = str.replaceAll("|","_"); // Noncompliant; yields _/_F_i_l_e_|_N_a_m_e_._t_x_t_
 String clean5 = str.replaceFirst("|","_"); // Noncompliant;
```
Comment (vilchik-elena) [inline rules/S2639/java/rule.adoc:21]:
```suggestion
String clean = str.replaceAll("|","_"); // Noncompliant; yields _/_F_i_l_e_|_N_a_m_e_._t_x_t_
String clean2 = str.replaceFirst("|","_"); // Noncompliant; 
```

File: rules/S2639/java/rule.adoc [hunk 7]
```diff
@@ -1,29 +1,40 @@
...
+Here are some common mistakes to avoid, as these expressions will not work as you intended.
 
-The following should not be used as regular expressions:
-
-* ``++.++`` - matches any single character. Used in ``++replaceAll++``, it matches _everything_
-* ``++|++`` - normally used as an option delimiter. Used stand-alone, it matches the space between characters
-* ``++File.separator++`` - matches the platform-specific file path delimiter. On Windows, this will be taken as an escape character
-
-=== Noncompliant code example
+* When you use ``.`` in ``replaceAll`` it matches _everything_.
```
Comment (vilchik-elena) [inline rules/S2639/java/rule.adoc:7]:
```suggestion
When you use ``.`` in ``replaceAll`` it matches _everything_.
```

Comment (vilchik-elena) [inline rules/S2639/java/rule.adoc:7]: `replaceAll` just an example in fact, there are many other functions that take regex as parameter.

File: rules/S2639/javascript/rule.adoc [hunk 8]
```diff
@@ -1,22 +1,20 @@
...
-[source,javascript]
-----
-/^foo/.test("foobar"); // true
-----
+== Resources
+
+=== Documentation
+
+- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions/Character_classes[MDN web docs: Character classes]
+- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions[MDN web docs: Regular exporessions]
```
Comment (vilchik-elena) [inline rules/S2639/javascript/rule.adoc:17]:
```suggestion
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions[MDN web docs: Regular expressions]
```

File: rules/S2639/java/rule.adoc [hunk 9]
```diff
@@ -1,29 +1,40 @@
...
+String clean2 = str.replaceAll("|","_"); // Noncompliant; yields _/_F_i_l_e_|_N_a_m_e_._t_x_t_
 String clean5 = str.replaceFirst("|","_"); // Noncompliant; 
+----
+
+When you use ``File.separator``, it matches the platform-specific file path delimiter. On Windows, this will be taken as an escape character.
+
+[source,java]
+----
+String clean3 = str.replaceAll(File.separator,""); // Noncompliant; exception on Windows
 String clean6 = str.replaceFirst(File.separator,""); // Noncompliant;
```
Comment (vilchik-elena) [inline rules/S2639/java/rule.adoc:29]:
```suggestion
String clean = str.replaceAll(File.separator, ""); // Noncompliant; exception on Windows
String clean2 = str.replaceFirst(File.separator, ""); // Noncompliant;
```

File: rules/S2639/java/rule.adoc [hunk 10]
```diff
@@ -1,29 +1,40 @@
...
+
+``|`` is the OR operator.  When you use ``|`` as a stand-alone, it matches the space between characters. To use it as a literal character in a regular expression, it needs to be escaped with a backslash ``\|``.
+
+[source,java]
+----
+String clean2 = str.replaceAll("|","_"); // Noncompliant; yields _/_F_i_l_e_|_N_a_m_e_._t_x_t_
 String clean5 = str.replaceFirst("|","_"); // Noncompliant; 
+----
+
+When you use ``File.separator``, it matches the platform-specific file path delimiter. On Windows, this will be taken as an escape character.
```
Comment (vilchik-elena) [inline rules/S2639/java/rule.adoc:24]: I am not sure what "escape character" is. Are you?

**Review by vilchik-elena — CHANGES_REQUESTED** [review]
Looks good for JS (just one typo).
I suggest reverting Java as clearly more work is required to clarify all the cases. (or you can open a separate PR and pass it to JVM squad to finish)

**Review by vilchik-elena — APPROVED** [review]

**Comment by vilchik-elena** [comment]
@cynthiabethea could you fix the PR title and merge?

**Pull request merged**